### PR TITLE
feat(iam): openid connect provider

### PR DESCRIPTION
**Pull request by eladb**
### Commit Message
feat(iam): openid connect providers

Implements `iam.OpenIdConnectProvider` through a custom resource.

See README for details.

Related #5388
Related #3949
Related #6308

### End Commit Message

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-iam/README.md [hunk 1]
```diff
@@ -238,6 +238,36 @@ const newPolicyDocument = PolicyDocument.fromJson(policyDocument);
 
 ```
 
+### OpenID Connect Providers
+
+OIDC identity providers are entities in IAM that describe an external identity
+provider (IdP) service that supports the [OpenID Connect] (OIDC) standard, such as
+Google or Salesforce. You use an IAM OIDC identity provider when you want to
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-iam/README.md:245]: No love for Cognito? :)

File: packages/@aws-cdk/aws-iam/lib/oidc-provider.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+  /**
+   * A list of server certificate thumbprints for the OpenID Connect (OIDC)
+   * identity provider's server certificates.
+   *
+   * Typically this list includes only one entry. However, IAM lets you have up
+   * to five thumbprints for an OIDC provider. This lets you maintain multiple
+   * thumbprints if the identity provider is rotating certificates.
+   *
+   * The server certificate thumbprint is the hex-encoded SHA-1 hash value of
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-iam/lib/oidc-provider.ts:60]: Uppercase or lowercase, if that matters?

File: packages/@aws-cdk/aws-iam/lib/oidc-provider.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+      },
+    });
+
+    this.openIdConnectProviderArn = Token.asString(resource.ref);
+  }
+
+  public get stack() { return Stack.of(this); }
+
+  private getOrCreateProvider() {
+    return CustomResourceProvider.getOrCreate(this, RESOURCE_TYPE, {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-iam/lib/oidc-provider.ts:136]: Timeout?

File: packages/@aws-cdk/aws-iam/README.md [hunk 4]
```diff
@@ -238,6 +238,36 @@ const newPolicyDocument = PolicyDocument.fromJson(policyDocument);
...
+
+[OpenID Connect]: http://openid.net/connect
+[About Web Identity Federation]: https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_providers_oidc.html
+
+The following examples defines an OpenID Connect provider. Two client IDs
+(audiences) are will be able to send authentication requests to
+https://openid/connect.
+
+```ts
+new OpenIdConnectProvider(this, 'MyProvider', {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-iam/README.md:260]: Any guidance on how we use this provider?

I know you're just building it for cdk8s but you seem to be implying we can tie this to AWS identities in some way...?

Comment (eladb) [reply packages/@aws-cdk/aws-iam/README.md:260]: I added an example on how to use with cognito's `CfnIdentityPool`. Once we have an L2, we will need to switch that up (copy @nija-at).

File: packages/@aws-cdk/aws-iam/lib/oidc-provider.ts (hunk 3 above)
Comment (eladb) [reply packages/@aws-cdk/aws-iam/lib/oidc-provider.ts:136]: The default timeout for CR providers is 15min.

File: packages/@aws-cdk/aws-iam/lib/oidc-provider.ts (hunk 2 above)
Comment (eladb) [reply packages/@aws-cdk/aws-iam/lib/oidc-provider.ts:60]: I don't know. This is basically a copy & pasta from AWS docs.

File: packages/@aws-cdk/aws-iam/README.md (hunk 1 above)
Comment (eladb) [reply packages/@aws-cdk/aws-iam/README.md:245]: added

**Pull request merged**